lib/service/rpmsg/rpc/rpmsg_rpc_server.c

```diff
@@ -0,0 +1,129 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <errno.h>
+#include <metal/mutex.h>
+#include <metal/spinlock.h>
+#include <metal/utilities.h>
+#include <openamp/open_amp.h>
+#include <openamp/rpmsg_rpc_client_server.h>
+#include <string.h>
+#include <fcntl.h>
+
+#define raw_printf(format, ...) printf(format, ##__VA_ARGS__)
+#define LPRINTF(format, ...) raw_printf("Master> " format, ##__VA_ARGS__)
+#define LPERROR(format, ...) LPRINTF("ERROR: " format, ##__VA_ARGS__)
+
+int request_termination;
+int ept_deleted;
+int err_cnt;
+static struct rpmsg_rpc_services *table[TABLE_SIZE];
+
+void *copy_from_shbuf(void *dst, void *shbuf, int len)
+{
+	void *retdst = dst;
+
+	while (len && ((uintptr_t)shbuf % sizeof(int))) {
+		*(uint8_t *)dst = *(uint8_t *)shbuf;
+		dst++;
+		shbuf++;
+		len--;
+	}
+	while (len >= (int)sizeof(int)) {
+		*(unsigned int *)dst = *(unsigned int *)shbuf;
+		dst += sizeof(int);
+		shbuf += sizeof(int);
+		len -= sizeof(int);
+	}
+	while (len > 0) {
+		*(uint8_t *)dst = *(uint8_t *)shbuf;
+		dst++;
+		shbuf++;
+		len--;
+	}
+	return retdst;
+}
+
+int linux_rpmsg_rpc_server_init(int id, int (*cb_function)
+				(void *, struct rpmsg_endpoint *))
+{
+	struct rpmsg_rpc_services *service =
+	(struct rpmsg_rpc_services *)malloc(sizeof(struct rpmsg_rpc_services));
+
+	service->id = id;
+	service->respcb.cb_function = cb_function;
+
+	int hash_index = id % TABLE_SIZE;
+
+	while (table[hash_index]) {
+		++hash_index;
+		hash_index %= TABLE_SIZE;
+	}
+
+	table[hash_index] = service;
+
+	return hash_index;
+}
+
+struct rpmsg_rpc_services *find_service(int id)
+{
+	int hash_index = id % TABLE_SIZE;
```

Comment:
I don't understand this function 
rpc id should be independent from the index of the table.
That said this function should be probably reworked based on previous comment concerning the management of the rpmsg_rpc_services table